### DISCUSSION TITLE
How to load pytorch model file directly?

### DISCUSSION THREAD

**Discussion by csxiang18** [deepjavalibrary/djl#408] [Q&A]
## Question

I've already export a pytorch model into a pt file. For example: `model/my_model.pt`, this directory is under the resources directory.

When I want to load this model, I use this way: 
    
    String modelPath = "model";
    URI modelUri = classLoader.getResource(modelPath).toURI();
    Path modelDir = Paths.get(modelUri);
    Model ptModel = Model.newInstance("my_model", Device.cpu());
    ptModel.load(modelDir);

Is there any other way that I could use like this, load the pt file directly: `ptModel.load("model/my_model.pt")`?

**Comment by lanking520** [comment]
Starting from 0.9.0, you can load from modelDir from ModelZoo:

https://github.com/awslabs/djl/blob/master/api/src/main/java/ai/djl/repository/zoo/Criteria.java#L343

You can refer to our doc here: http://docs.djl.ai/master/docs/load_model.html. There are many ways to load: like http, hdfs, s3 or simply local file paths are all supported.

If you are looking for loading the model directly, you can do the followings:
```
ptModel.load(Paths.get("model"), "my_model");
```
Which is very similar when you create new instance of the model. There seemed to be no way to bypass setting model name from the `Model` API. But doable if you try to load from the `ZooModel` api

**Reply by dtneverdie** [reply]
I am using the way you mentioned of 'model.load' to load a local pytorch model now. Here comes my code:`
Path modelDir = Paths.get("resources");//checkpoint.pt is in /src/resourses
Model ptModel = Model.newInstance("checkpoint", Device.cpu());
ptModel.load(modelDir, "checkpoint");`
And exception information is`Exception in thread "main" java.io.FileNotFoundException: Parameter file with prefix: resources not found in: D:\DeepJavaLibraryExample-master\resources
	at ai.djl.mxnet.engine.MxModel.load(MxModel.java:98)
	at ai.djl.Model.load(Model.java:121)
	at mxnet.inference.AnotherTest.main(AnotherTest.java:34)`
Now i get stuck in this problem and dont know how to load it

**Reply by frankfliu** [reply]
@dtneverdie 
We recommend you to use Criteria API, it's more portable and easy to change the way you deploy your model in future (.e.g, change engine, path, GPU, etc)

The exception you got is because you included both MXNet and PyTorch engine in your project. DJL try to use MXNet to load model first. There are multiple way to resolve this issue:
1. Remove MXNet engine from your project's dependency
2. or set PyTorch as your default engine: `System.setProperties("ai.djl.default_engine", "PyTorch");`
3. Create PyTorch model explicitly: `Model model.= Model.newInstance("checkpoint", Device.cpu(), "PyTorch");`
4. If you use Criteria API, add `.optEngine("PyTorch")` to criteria

**Reply by dtneverdie** [reply]
Thx a lot, but now it shows new exception:
`ai.djl.engine.EngineException: Failed to load PyTorch native library
	at ai.djl.pytorch.engine.PtEngine.newInstance(PtEngine.java:77)
	at ai.djl.pytorch.engine.PtEngineProvider.getEngine(PtEngineProvider.java:40)
	at ai.djl.engine.Engine.getEngine(Engine.java:161)
	at ai.djl.Model.newInstance(Model.java:98)
	at ai.djl.repository.zoo.BaseModelLoader.createModel(BaseModelLoader.java:183)
	at ai.djl.repository.zoo.BaseModelLoader.loadModel(BaseModelLoader.java:145)
	at ai.djl.repository.zoo.Criteria.loadModel(Criteria.java:166)
	at ai.djl.repository.zoo.ModelZoo.loadModel(ModelZoo.java:123)
	at mxnet.inference.TestingModel.loadDefaultDnnModel(TestingModel.java:52)
	at mxnet.inference.TestingModel.main(TestingModel.java:62)
Caused by: java.lang.IllegalStateException: Native library not found
	at ai.djl.pytorch.jni.LibUtils.getLibName(LibUtils.java:88)
	at ai.djl.pytorch.jni.LibUtils.loadLibrary(LibUtils.java:72)
	at ai.djl.pytorch.engine.PtEngine.newInstance(PtEngine.java:50)
	... 9 more
`
I have tried both Criteria `Criteria<Float[], Float[]> criteria = Criteria.builder()
                    .setTypes(Float[].class, Float[].class)
                    .optTranslator(translator)
                    .optModelUrls("jar:///checkpoint_gsn_latest.zip")
                    .optModelName("checkpoint_gsn_latest")
                    .optProgress(new ProgressBar())
                    .optEngine("PyTorch")
                    .build();
            ZooModel<Float[], Float[]> model = ModelZoo.loadModel(criteria);`and 
`Model ptModel = Model.newInstance("checkpoint_gsn_latest", Device.cpu(), "PyTorch");
        //ptModel.load(p, "checkpoint_gsn_latest");
        ptModel.load(Paths.get("resources"), "checkpoint_gsn_latest.pt");`
 model.newInstance ways, and exception information is the same

**Reply by lanking520** [reply]
Looks like the native library loading is failed. This is nothing related to model itself. Which OS you are on? And which DJL and PyTorch version you are using?

**Reply by frankfliu** [reply]
Did you include pytorch native library in your project?

**Reply by dtneverdie** [reply]
> Did you include pytorch native library in your project?

I have added pytorch-native-auto to my dependence. If you mean the native library part in http://docs.djl.ai/pytorch/pytorch-native/index.html, this gradlew commonds dont work on my maven project

**Reply by dtneverdie** [reply]
i haved tried to use gradlew compileJNI on a new gradle project and the exception sayed `Task 'compileJNI' not found in root project 'DJLTest'.
`

**Reply by dtneverdie** [reply]
> Looks like the native library loading is failed. This is nothing related to model itself. Which OS you are on? And which DJL and PyTorch version you are using?

win10, djl version is 0.13.0-SNAPSHOT,pytorch-native-auto:1.8.1

**Reply by frankfliu** [reply]
Can you try examples in DJL repository directly?
```
cd djl/pytorch/pytorch-model-zoo
gradlew test
```

**Reply by dtneverdie** [reply]
> Can you try examples in DJL repository directly?
> 
> ```
> cd djl/pytorch/pytorch-model-zoo
> gradlew test
> ```

There is no gradlew.bat in this directory so i used `..\..\ gradlew test ` command instead and it showed that `BUILD SUCCESSFUL in 12s
9 actionable tasks: 9 executed`.
Well i ran my project again and exceptions still exist. Failed to load Pytorch native library problem still the same but now it is caused by 
`Caused by: java.lang.UnsatisfiedLinkError: C:\Users\57495\.djl.ai\pytorch\1.8.1-cu102-win-x86_64\0.13.0-SNAPSHOT-cu102-djl_torch.dll: **找不到指定的程序**。
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1086)
	at ai.djl.pytorch.jni.LibUtils.loadNativeLibrary(LibUtils.java:351)
	at ai.djl.pytorch.jni.LibUtils.loadLibrary(LibUtils.java:77)
	at ai.djl.pytorch.engine.PtEngine.newInstance(PtEngine.java:50)
	... 4 more。` The Chinese words mean 'cant find appointed application'. I dont know why this error came in local language.

**Reply by frankfliu** [reply]
DJL 0.13.0-SNAPSHOT should use pytorch-native-auto:1.9.0-SNAPSHOT

You can use  DJL 0.12.0 if use pytorch-native-auto:1.8.1

**Reply by dtneverdie** [reply]
> DJL 0.13.0-SNAPSHOT should use pytorch-native-auto:1.9.0-SNAPSHOT
> 
> You can use DJL 0.12.0 if use pytorch-native-auto:1.8.1

now it shows `ai.djl.engine.EngineException: `torch::jit::load()` received a file from `torch.save()`, but `torch::jit::load()` can only load files produced by `torch.jit.save()`
	at ai.djl.pytorch.jni.PyTorchLibrary.moduleLoad(Native Method)
	at ai.djl.pytorch.jni.JniUtils.loadModule(JniUtils.java:1359)
	at ai.djl.pytorch.engine.PtModel.load(PtModel.java:85)
	at ai.djl.repository.zoo.BaseModelLoader.loadModel(BaseModelLoader.java:152)
	at ai.djl.repository.zoo.Criteria.loadModel(Criteria.java:166)
	at ai.djl.repository.zoo.ModelZoo.loadModel(ModelZoo.java:123)
	at mxnet.inference.TestingModel.loadDefaultDnnModel(TestingModel.java:52)
	at mxnet.inference.TestingModel.main(TestingModel.java:62)`

**Reply by frankfliu** [reply]
@dtneverdie 
As the error message state, your model is not jit traced model. see: http://docs.djl.ai/master/docs/pytorch/how_to_convert_your_model_to_torchscript.html

**Comment by frankfliu** [comment]
@csxiang18 If you put model in resource file, most likely you will pack the model in a jar file when deploy it.

I this case, you have to use jar:// url to load the model: http://docs.djl.ai/docs/load_model.html#load-model-from-a-url

```
Criteria<Image, Classifications> criteria = Criteria.builder()
        .setTypes(Image.class, Classifications.class) // defines input and output data type
        .optTranslator(ImageClassificationTranslator.builder().setSynsetArtifactName("synset.txt").build())
        .optModelUrls("jar:///model")
        .optEngine("PyTorch")
        .optModelName("my_model")
        .build();

ZooModel<Image, Classifications> model = criteria.loadModel();
```

**Comment by csxiang18** [comment]
Thanks for answering this question, as it's my first time to try DJL, so may some API I do not use very correctly. In fact I want to pack the model in the jar file as a lib to deploy so that others could use this. Maybe I could state the problem more clearly. 
@frankfliu my model is a self-defined model, for-example, a 3 layer dnn, no existing model in ModelZoo. I follow this doc: [Export Pytorch Model](https://djl.ai/docs/pytorch/how_to_convert_your_model_to_torchscript.html) to export the model as a my_model.pt file. Then I put the model file under the resource directory `model/my_model.pt` so that I could use getResouce to load the model. 
```
ClassLoader classLoader = LoadModelTest.class.getClassLoader();
String modelPath = "model";
URI modelUri = classLoader.getResource(modelPath).toURI();
```
Then I want to provide a lib, and write like this, so others could use the function like this: `LoadModelTest.predict(features)`

```
package com.test;

public class LoadModelTest {
    public static Model model;
    static {
        try {
            loadModel();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    public static void main(String[] args) {
        float[] features = new float[199];
        float[] output = predict(features);
        System.out.println(output[0]);
    }

    public static float[] predict(float[] features) {
        try {
            Predictor<float[], float[]> predictor = model.newPredictor(new FloatToFloatTranslator());
            return predictor.predict(features);

        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static void loadModel() throws Exception {
        ClassLoader classLoader = LoadModelTest.class.getClassLoader();
        String modelPath = "model";
        URI modelUri = classLoader.getResource(modelPath).toURI();
        Path modelDir = Paths.get(modelUri);
        model = Model.newInstance("my_model", Device.cpu());
        model.load(modelDir);
    }

    private static final class FloatToFloatTranslator implements Translator<float[], float[]> {

        @Override
        public Batchifier getBatchifier() {
            return Batchifier.STACK;
        }

        @Override
        public float[] processOutput(TranslatorContext translatorContext, NDList ndList) throws Exception {
            NDArray tempArray = ndList.get(0);
            return tempArray.toFloatArray();
        }

        @Override
        public NDList processInput(TranslatorContext translatorContext, float[] features) throws Exception {
            NDManager manager = translatorContext.getNDManager();
            return new NDList(manager.create(features));
        }
    }
}

```

The above script works fine when I wrote the lib. However, when I package this lib, it could not load the model correctly. As `my_model.pt` is in the Jar file and Paths could not parse the model directory correctly. So how could I solve this? 

Another question is that in the doc, it said that the predictor is not thread-safe. So the above way is the correct way to predict features?

**Comment by frankfliu** [comment]
@csxiang18 
The PyTorch engine can only load model from file system. So the `loadModel()` can only take `Path`

If you package your classes in a jar file, the URL returned from `classLoader.getResource` is a special URL that only supported by JVM something like: "jar://your.jar!/model". `Path modelDir = Paths.get(modelUri);` won't give you a correct file system path, so load model will fail since Path is not valid.

To resolve this issue, you can use Criteria API to load your model as I posted earlier. The api can take care both jar and unjared case.

```
Criteria<Float[], Float[]> criteria = Criteria.builder()
        .setTypes(Float[].class, Float[].class) // defines input and output data type
        .optTranslator(new FloatToFloatTranslator())
        .optModelUrls("jar:///model")
        .optModelName("my_model")
        .build();

ZooModel<Float[], Float[]> model = ModelZoo.loadModel(criteria);
```
